In [ ]:
# Import packages
import pandas as pd 
import numpy as np 

## Q1. Pandas version

In [148]:
pd.__version__

'3.0.6'

**Ans: the version of pandas is `3.0.6`**

----

### Getting the data

In [149]:
df = pd.read_csv("car_fuel_efficiency_2026.csv")
df.head(5)

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


----
## Q2. Records count
How many records are in the dataset?

In [150]:
df.shape

(10000, 11)

In [151]:
df.shape[0]

10000

**Ans: the total records equal to the result from `df.shape[0]`, which is 10,000 records**

----

## Q3. Fuel types
How many fuel types are presented in the dataset?

In [152]:
df['fuel_type'].nunique()

3

In [153]:
df['fuel_type'].describe()

count        10000
unique           3
top       Gasoline
freq          6913
Name: fuel_type, dtype: object

**Ans: there are 3 fuel types presented in the dataset**

----

## Q4. Missing Values
How many columns in the dataset have missing values?

In [154]:
df.isnull().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

**Ans: According to the result above, there are two columns contain missing values**

----

## Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [155]:
df.groupby('origin').fuel_efficiency_mpg.max()

origin
Asia      41.2
Europe    39.4
USA       38.2
Name: fuel_efficiency_mpg, dtype: float64

**Ans: the maximum fuel efficiency of cars from Asia is 41.2**

----

## Q6. Median value of horsepower

1. Find the median value of the `horsepower` column in the dataset
2. Next, calculate the most frequent value of the same `horsepower` column
3. Use the `fillna` method to fill the missing values in the `horsepower` column with the most frequent value from the previous step
4. Now, calculate the median value of `horsepower` once again

Has it changed?

In [156]:
# Step 1
df['horsepower'].median()

np.float64(254.0)

the median of `horsepower` is 254

In [157]:
# Step 2
df['horsepower'].mode()

0    252.0
Name: horsepower, dtype: float64

the most frequent value of `horsepower` is 252

In [158]:
# Step 3
mode_val = float(df['horsepower'].mode().iloc[0])
print(mode_val)
df['horsepower'] = df['horsepower'].fillna(mode_val)

252.0


In [159]:
# step 4
df['horsepower'].median()

np.float64(252.0)

**Ans: Yes, it decreased.**

**We can see that the median of `horsepower` decreased from 254 to 252, after we filled the missing values with the most frequent value of `horsepower`: 252**

----

## Q7. Sum of weights
1. Select all the cars from Asia
2. Select only columns `vehicle_weight` and `model_year`
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the sum of all the elements of the result?

In [160]:
# step 1 - 3
df_asia = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7)
df_asia

,vehicle_weight,model_year
2,4240,1996
6,4310,1992
8,4230,1997
10,4420,1978
15,3950,2007
16,4750,2021
22,4450,2019


In [161]:
# step 4
X = np.array(df_asia)
X

array([[4240, 1996],
       [4310, 1992],
       [4230, 1997],
       [4420, 1978],
       [3950, 2007],
       [4750, 2021],
       [4450, 2019]])

In [162]:
# step 5 - 6
XTX = X.T.dot(X)
XTX_inv = np.linalg.inv(XTX)
print(XTX_inv)

[[ 2.96830537e-06 -6.43071025e-06]
 [-6.43071025e-06  1.39675281e-05]]


In [164]:
# step 7
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [ ]:
# step 8 - 9
temp = XTX_inv.dot(X.T)
w = temp.dot(y)
w.sum()

np.float64(0.3691969690492751)

**Ans: the sum of w is 0.369**

----
End of this page